proxy/ipvs: Only compute node ip addresses once per sync

cezarsa opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

This PR rearranges the `syncProxyRules` loop when kube-proxy is using IPVS to only compute node IPs once per sync.

Previously the same ip addresses would be computed for each nodePort service and this could be CPU intensive for a large number of nodePort services with a large number of ipaddresses on the node.

**Which issue(s) this PR fixes**:
Possibly related to #73382.

**Special notes for your reviewer**:

cc @andrewsykim
/sig network

**Does this PR introduce a user-facing change?**:
```release-note
Reduce kube-proxy cpu usage in IPVS mode when a large number of nodePort services exist.
```

cezarsa committed:
proxy/ipvs: Only compute node ip addresses once per sync

Previously the same ip addresses would be computed for each nodePort
service and this could be CPU intensive for a large number of nodePort
services with a large number of ipaddresses on the node.

tedyu commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,42 @@ func (proxier *Proxier) syncProxyRules() {
+	if hasNodePort {
+		nodeAddresses, err = utilproxy.GetNodeAddresses(proxier.nodePortAddresses, proxier.networkInterfacer)
+		if err != nil {
+			klog.Errorf("Failed to get node ip address matching nodeport cidr: %v", err)
```
why not to skip iterating nodeAddresses in this case ?

cezarsa replied:
This is what will happen in practice. When `GetNodeAddresses` returns an error it also returns a nil value for `nodeAddresses` and this will cause the iteration to be skipped.

andrewsykim replied:
This isn't directly obviously though, I agree with tedyu it's easier to follow with an explicity skip or check. You can use `nodeAddresses.Len()` to do an explicity check before iterating node addresses

cezarsa replied:
Done.

andrewsykim commented:
/assign

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,42 @@ func (proxier *Proxier) syncProxyRules() {
+		if err != nil {
+			klog.Errorf("Failed to get node ip address matching nodeport cidr: %v", err)
+		}
+		for address := range nodeAddresses {
```
use `nodeAddresses.List()` here

cezarsa replied:
Done, I actually changed `nodeAddresses` to a `[]string` that is only set if `nodeAddrSet.Len() > 0` to avoid multiple calls to `.List()` down the road.

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,42 @@ func (proxier *Proxier) syncProxyRules() {
+				if err != nil {
+					klog.Errorf("Failed to list all node IPs from host, err: %v", err)
+				}
+				break
```
Is this `break` intentional?

cezarsa replied:
Yes, that's intentional. The reasoning was that if `IsZeroCIDR` is true to any address then `nodeIPs` will be set to all host IPs (the return of `ipGetter.NodeIPs()`) and we don't need to append extra IP addresses to `nodeIPs` and it's safe to break here.

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1064,14 +1100,14 @@ func (proxier *Proxier) syncProxyRules() {
 
 			var lps []utilproxy.LocalPort
-			for address := range addresses {
+			for address := range nodeAddresses {
```
nodeAddresses.List()

andrewsykim commented:
Thanks for the PR @cezarsa! 

/ok-to-test

lbernail commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,45 @@ func (proxier *Proxier) syncProxyRules() {
+		if nodeAddrSet.Len() > 0 {
+			nodeAddresses = nodeAddrSet.List()
+			for _, address := range nodeAddresses {
+				if utilproxy.IsZeroCIDR(address) {
```
Is there a case where `nodeAddresses` could contain 0.0.0.0/0 and other addresses?
Because we could
- if `nodeAddrSet.Len()==1` and nodeAddrSet[0]=`0.0.0.0/0` => iterate over `nodeAddrSet`
- otherwise iterate over `proxier.ipGetter.NodeIPs()`

cezarsa replied:
Assuming the node has three interfaces:
```
eth0:    10.0.0.1, fe80::1c00:a8ff:fe00:481f
lo0:     127.0.0.1, ::1
docker0: 172.17.0.1
```

It's possible for nodeAddresses to contain both `::/0` and another IP, but
only when mixing IPv4 and IPv6 arguments due to checks here:

https://github.com/kubernetes/kubernetes/blob/289430a83ebf9c637b9782492aa038e029ee80c5/pkg/proxy/util/utils.go#L179-L184

Starting kube-proxy with `--nodeport-addresses=::/0,10.0.0.1/32` would cause these values:

```
nodeAddresses == ["10.0.0.1", "::/0"]                    # This is the order sets.String.List would return them.
nodeIPs       == ["10.0.0.1", "127.0.0.1", "172.17.0.1]  # IPv6 address possibly here also, if there are routes using them as source.
```

Down the line, when configuring a node port service we would have:

```
lps == []utilproxy.LocalPort{
    {
        IP: "10.0.0.1",
        ...
    },
    {
        IP: "",
    },
}
```

I'm not sure this value for `lps` makes much sense but I didn't touch this part of the code. If it's okay for it to have a single empty IP entry then I guess `utilproxy.GetNodeAddresses` could be modified to return only `::/0` and them skip iterating on `nodeAddresses` if `utilproxy.IsZeroCIDR(nodeAddresses[0])` is true.

andrewsykim commented:
/priority important-soon

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,45 @@ func (proxier *Proxier) syncProxyRules() {
+		}
+		if nodeAddrSet.Len() > 0 {
+			nodeAddresses = nodeAddrSet.List()
+			for _, address := range nodeAddresses {
```
For this PR, can we keep the logic the same as what it was before and open a separate PR for the zero cidr changes? 

i.e. keep this section as:
```go
                      for _, address := range nodeAddresses {
				if !utilproxy.IsZeroCIDR(address) {
					nodeIPs = append(nodeIPs, net.ParseIP(address))
					continue
				}
				// zero cidr
				nodeIPs, err = proxier.ipGetter.NodeIPs()
				if err != nil {
					klog.Errorf("Failed to list all node IPs from host, err: %v", err)
				}
			}
```

andrewsykim replied:
Fwiw I think your current changes do make more sense since we should break on zero cidr but easier to isolate the PRs that way

andrewsykim replied:
Not opposed to separate commits either actually

cezarsa replied:
Thanks for the feedback @andrewsykim! I have updated the PR separating the changes in 2 different commits and trying to explain the reasoning better in the commit messages.

andrewsykim replied:
thank you!

cezarsa committed:
proxy/ipvs: Compute all node ips only once when a zero cidr is used

Computing all node ips twice would always happen when no node port
addresses were explicitly set. The GetNodeAddresses call would return
two zero cidrs (ipv4 and ipv6) and we would then retrieve all node IPs
twice because the loop wouldn't break after the first time.

Also, it is possible for the user to set explicit node port addresses
including both a zero and a non-zero cidr, but this wouldn't make sense
for nodeIPs since the zero cidr would already cause nodeIPs to include
all IPs on the node.

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,46 @@ func (proxier *Proxier) syncProxyRules() {
+		if err != nil {
+			klog.Errorf("Failed to get node ip address matching nodeport cidr: %v", err)
+		}
+		if nodeAddrSet.Len() > 0 {
```
This will panic if err != nil above

cezarsa replied:
Actually this works because the set implementation will simply call len() on the underlying map which will return 0 even for a nil map: https://play.golang.org/p/ZO4U9-VvxPs. I agree this is confusing though, I'll update this condition to `if err == nil && nodeAddrSet.Len() > 0`

andrewsykim replied:
Good point though on the nil map

andrewsykim replied:
Good point though on the nil map

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -808,6 +808,46 @@ func (proxier *Proxier) syncProxyRules() {
+		// List of node addresses to listen on if a nodePort is set.
+		nodeAddresses []string
+		// List of node IP addresses to be used as IPVS services if nodePort is
+		// set.
```
small nit: make this comment into a single line?

cezarsa replied:
Done.

andrewsykim reviewed (commented):
/approve
/assign @lbernail 

One really minor comment, assigning @lbernail for final lgtm

andrewsykim commented:
@cezarsa can you update the release notes please?

lbernail commented:
/lgtm

Thanks again!

Nice improvement for later: modify `GetNodeAddresses` and `realIPGetter.NodeIPs()` to ignore kube-ivps0 interface (even if it runs only once now, going through possibly 1000s unnecessary IPs will take some time that we can easily avoid)

lbernail commented:
We should cherry-pick this fix to all supported releases

cezarsa commented:
Thanks @lbernail! About your suggestion on ignoring the `kube-ipvs0` interface. I've been taking a look at this, and by no means I'm anywhere near to fully understanding how this works on the kernel side, but from what I could gather there's no feasible way to filter which interfaces we want the kernel to list addresses or routes.

To retrieve the IP addresses from the kernel a netlink [RTM_GETADDR](http://man7.org/linux/man-pages/man7/rtnetlink.7.html) request is issued. One of the flags that can be be set is [NLM_F_MATCH](http://man7.org/linux/man-pages/man7/netlink.7.html) to filter the response, however according to the man page this flag is not implemented yet in the kernel. This also applies to getting routes from the kernel with [RTM_GETROUTE](http://man7.org/linux/man-pages/man7/rtnetlink.7.html). I found this somewhat old discussion on the subject here https://lists.infradead.org/pipermail/libnl/2013-June/001014.html.
Also, a patch to handle NLM_F_MATCH on ipv6 routes was rejected in https://patchwork.ozlabs.org/patch/133440/

To make sure this is what happens today I tested running `strace` on this snippet https://play.golang.org/p/vtBjdRhwO3K which should only list addresses for the `lo` interface and also on `ip addr list lo`:

```
strace -e trace=network ip addr list lo
```

On both cases I can see on the response messages all interfaces addresses being returned by the kernel (strace output edited for clarity):

```
sendto(3 ... type=RTM_GETADDR
...
recvmsg(3 ... if_nametoindex("lo") ... if_nametoindex("enp0s3") ... if_nametoindex("enp0s8")
```

Sure we could still filter try improving the way we filter the `kube-ivps0` interface on our side, however I'm not sure it would help a lot since the kernel would still be doing most of the work of enumerating all ip addresses on it anyway.

Again I might be wrong as I have very little experience with interacting directly with netlink. It would be great if some of you folks with more kernel/netlink expertise could take a look at this and see if this makes sense.

mrrandrade commented:
Hey guys!

Original bug poster here, I've come here just to give a warm thank you to all involved with this.

I wanted to post the "benchmarks" of the several improvements made to kube-proxy on my setup.

My setup has around 20k lines of ipvsadm -Ln. I have a machine on my cluster running exactly 0 workloads - no pod workload, basically nothing other than kubernetes/calico using CPU.

These are the timings:

Old 1.12.2 kube-proxy: 221s to get from 0 to 19439 lines of ipvsadm;
New 1.16.2 kube-proxy: 4s to get from 0 to 19438 lines of ipvsadm.

Thank you very much!
